Publish index and archive actions to SNS topic

Ariana-B opened the pull request:
In response to [issue #516](https://github.com/opendatacube/odc-tools/issues/516), publishing indexing or archiving action to an SNS topic.

- Added a flag `--publish-action [sns topic arn]` which publishes action ("ARCHIVED"/"ADDED") and STAC document to the SNS topic when set. 
- SNS publication function defined in queue lib since it is similar/related to other SQS functionality (and wasn't sure where else would be better).
- Wrote function `ds_to_stac` to convert from ds eo3 metadata to STAC
- Updated dc-tools readme to include `archive-less-mature` and `publish-action` flags
- Added a script that sets up the testing database
- Updated test instructions in README and added tests for sns publication

Commit:
sns publishing action

Commit:
remove skip import

jmettes commented on libs/cloud/odc/aws/queue.py:
```diff
@@ -180,3 +184,16 @@ def get_messages(
+    Publish a message to an sns topic
+    """
+    sns = boto3.resource("sns")
+    sns_topic = sns.create_topic(Name=SNS_TOPIC_NAME)
```
We'll have a topic already existing - `dea-public-data-sentinel-2`, and will want the tool to accept the topic name (or ARN [the Absolute Resource Name in AWS]) as a parameter - e.g., `--publish arn:aws:sns:ap-southeast-2:xxxxxxxxxxxxxxxxx:dea-public-data-sentinel-2` (where the xxx's are the account number, which I've forgotten)

jmettes commented on libs/cloud/odc/aws/queue.py:
```diff
@@ -180,3 +184,16 @@ def get_messages(
+
+    stac = json.dumps(stac)
+    sns_topic.publish(
+        Message='{"action": {action}, "stac": {stac}}',
```
I think you probably want

```
message_attributes = {
  "action": action,
   ... # more attributes? e.g., product_name?
}
sns_topic.publish(Message=stac, MessageAttributes=message_attributes)
```

jmettes replied:
prob enough to use [SNS client](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sns.html#SNS.Client.publish)

jmettes commented:
Awesome, thanks so much for working on this!

Commit:
get sns topic from publish-action flag instead of hardcoded; update param type accordingly and fix missing logic

Commit:
added tests

Commit:
add flag to readme

Commit:
make tests a little more robust

jmettes commented on libs/cloud/odc/aws/queue.py:
```diff
@@ -180,3 +181,30 @@ def get_messages(
+    Publish 'added' or 'archived' action to the provided sns topic
+    """
+    sns = boto3.client("sns")
+    sns_topic = sns.create_topic(Name=topic_name)
```
It's not really the job of these tools to create infrastructure like SNS topics - we set all that up elsewhere. I think this code should accept an already existing topic (i.e., the ARN), and not create it.

Ariana-B replied:
`create_topic` only creates a topic if it doesn't exist, otherwise it simply returns the arn. This is the only way to get a topic arn based off the name, though I can change the flag to take in the arn instead if that's preferable.

jmettes replied:
Oh I see. I don't think `odc-tools` should create infra. Rather do that explicitly elsewhere. I think ARN is fine

Ariana-B replied:
Done

jmettes commented on libs/cloud/odc/aws/queue.py:
```diff
@@ -180,3 +181,30 @@ def get_messages(
+    sns.publish(
+        TopicArn=sns_topic.get("TopicArn"),
+        Message=json.dumps(message),
+        MessageAttributes={
```
Is it worth adding more `MessageAttributes` such as the `product` (so we can filter on it, beyond just whether it was added/archived)? Maybe `maturity` is worth adding too?

Ariana-B replied:
What would the expected behaviour be for products that don't have `maturity`?

jmettes replied:
Not sure, haven't come across that. I'd guess don't pass it as a `MessageAttribute` if it's not in the STAC document.

jmettes commented on libs/cloud/odc/aws/queue.py:
```diff
@@ -180,3 +181,30 @@ def get_messages(
+    }
+    sns.publish(
+        TopicArn=sns_topic.get("TopicArn"),
+        Message=json.dumps(message),
```
The `Message` payload is just the STAC document itself. The `MessageAttributes` are used to describe it, so they can be filtered on when receiving lots of messages.

Our convention is as follows:

```
sns.publish(Message=stac_document, MessageAttributes={
    "action": { "DataType": "String", "StringValue": action },
    "product": {
        "DataType": "String", 
        "StringValue": stac_document["odc:product"] if "odc:product" in stac_doc else stac_document["collection"]
    },
    "maturity": {
        "DataType": "String",
        "StringValue": stac_doc["properties"]["dea:dataset_maturity"]
    }
})
```

Other message attributes I've come across are: `bbox.ll_lon`, `bbox.ll_lat`, `bbox.ur_lon`, `bbox.ur_lat`, `datetime`, `cloudcover`.

Ariana-B replied:
Changed

jmettes commented on libs/cloud/odc/aws/queue.py:
```diff
@@ -180,3 +181,30 @@ def get_messages(
+                "DataType": "String",
+                "StringValue": "ARCHIVED {or} ADDED",
+            },
+            "stac": {
```
We don't want to include the STAC doc inside the message attributes - that's the message payload

Ariana-B replied:
Fixed

jmettes commented on libs/cloud/odc/aws/queue.py:
```diff
@@ -180,3 +181,30 @@ def get_messages(
+        MessageAttributes={
+            "action": {
+                "DataType": "String",
+                "StringValue": "ARCHIVED {or} ADDED",
```
You'll want to pass the `action` variable value here

Ariana-B replied:
Done

jmettes commented on apps/dc_tools/README.md:
```diff
@@ -126,6 +126,12 @@ Options:
+  --archive-less-mature           Find less mature versions of the dataset and
+                                  archive them
+                                  
+  --publish-action TEXT           Publish indexing action to SNS topic
```
Maybe `TEXT` be `SNS topic ARN` for clarity

Ariana-B replied:
Done

Commit:
fix message attributes, use topic arn instead of name, update tests

Commit:
use arn in tests

jmettes commented on apps/dc_tools/README.md:
```diff
@@ -178,6 +184,11 @@ Options:
+  --archive-less-mature           Find less mature versions of the dataset and
+                                  archive them
+
+  --publish-action TEXT           Publish indexing action to SNS topic
```
SNS ARN here too?

Commit:
update readme

jmettes commented on apps/dc_tools/tests/test_sns_publishing.py:
```diff
@@ -0,0 +1,261 @@
+
+@mock_sns
+@mock_sqs
+def test_sqs_publishing_archive(aws_credentials, aws_env, sqs_message):
```
The only test case this doesn't cover is the scenario of using `--archive-less-mature`  + `--publish-action`. Might be worth adding, because it's the main thing I'll immediately be using it for

jmettes replied:
Or is this essentially covered via functionality of `---archive`?

Ariana-B replied:
Added

Commit:
added pytest fixtures and test for sns with archive-less-mature; readme instruction for adding new metadata types

jmettes reviewed (approved):
Woohoo! tyvm @Ariana-B - this looks fantastic. well done navigating the jungle that is odc-tools

Commit:
add eodatasets3 dependency to test env

Commit:
fix missing params, update product name

Commit:
fix missing params, update product name

Commit:
revert name change

Commit:
add sentinel-2-l2a to list of collections

Commit:
add s3 exception for empty doc stream, changed s2_l2a product definition in line with renaming

Commit:
update s2 l2a stac doc

Commit:
add script to setup database for testing

omad commented on apps/dc_tools/tests/test_sns_publishing.py:
```diff
@@ -0,0 +1,354 @@
+    aws_credentials, aws_env, odc_db_for_sns, s2am_dsid
+):
+    # set up sns topic and the sqs queue that is subscribed to it
+    sns = boto3.client("sns")
+    sqs = boto3.client("sqs")
+    topic_name = "test-topic"
+    queue_name = "test-queue"
+
+    topic = sns.create_topic(Name=topic_name)
+    queue = sqs.create_queue(QueueName=queue_name)
+    attrs = sqs.get_queue_attributes(
+        QueueUrl=queue.get("QueueUrl"), AttributeNames=["All"]
+    )
+    queue_arn = attrs["Attributes"]["QueueArn"]
+    sns_arn = topic.get("TopicArn")
+    sns.subscribe(
+        TopicArn=sns_arn,
+        Protocol="sqs",
+        Endpoint=queue_arn,
+    )
```
These lines could go nicely into a fixture, the fixture itself could be annotated with `@mock_sns` and `@mock_sqs`, and could provide the SNS ARN and a [boto3 SQS Queue](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs.html#queue) ready to receive messages from.

This would reduce the duplication and make the purpose of the tests easier to follow.

Ariana-B replied:
Done

omad commented on apps/dc_tools/odc/apps/dc_tools/_stac.py:
```diff
@@ -404,3 +413,21 @@ def stac_transform(input_stac: Document, relative: bool = True) -> Document:
+        stac = to_stac_item(
+            from_doc(ds.metadata_doc, skip_validation=True),
+            location,
+            explorer_base_url="https://explorer.dea.ga.gov.au",
```
If this is optional, we should leave it out instead of defaulting to DEA. We're not the only ones to use these tools.

Ariana-B replied:
Removed

omad commented on tests/test-env-py38.yml:
```diff
@@ -32,6 +32,7 @@ dependencies:
   - lxml # needed for thredds-crawler
   - urlpath
   - datadog
+  - eodatasets3
```
Can you please add this new dependency into `apps/dc_tools/setup.cfg` too. 🙏

Ariana-B replied:
Done

omad reviewed (changes requested):
Looks great Ariana. Lets try to merge this today before I get too stuck into updating tests and linting errors (conflict hell).
I love the tests!

Commit:
address comments

omad reviewed (approved):
Yay! Thanks!

omad committed:
Fix some failing tests

- include eodatasets3 in requirements.txt
- rm some bogus fs_to_dc tests
- revert reordering imports in odc.ui.__init__

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

omad commented:
I've rebased these changes and force-pushed them, but there are still some failing tests. Some are due to internal stuff, some due to external data changes.

I'm going to merge this, and keep working on fixes in a new branch and PR.
